vernac/declare.ml

```diff
@@ -1627,20 +1643,23 @@ let prepare_proof ~unsafe_typ { proof } =
       Vars.universes_of_constr t, t
     else to_constr_body t
   in
-  (* ppedrot: FIXME, this is surely wrong. There is no reason to duplicate
-     side-effects... This may explain why one need to uniquize side-effects
-     thereafter... *)
-  (* EJGA: actually side-effects de-duplication and this codepath is
-     unrelated. Duplicated side-effects arise from incorrect scheme
-     generation code, the main bulk of it was mostly fixed by #9836
-     but duplication can still happen because of rewriting schemes I
-     think; however the code below is mostly untested, the only
-     code-paths that generate several proof entries are derive and
-     equations and so far there is no code in the CI that will
-     actually call those and do a side-effect, TTBOMK *)
-  (* EJGA: likely the right solution is to attach side effects to the first constant only? *)
-  let proofs = List.map (fun (_, body, typ) -> (to_constr_body body, eff), to_constr_typ typ) initial_goals in
-  proofs, Evd.evar_universe_context evd
+  let make_proof eff (_, body, typ) =
+      (to_constr_body body, eff), to_constr_typ typ in
+  let regular_preparation initial_goals =
+    let fst_eff, more_eff =
```

Comment:
```suggestion
    (* When a bunch of constants are generated from the same proof session effects;
        - are attached to the first one if the constants are transparent
        - are attached to each constants if the constants are opaque
       This is because effects are not inlined in transparent constants (so the effects
       of the first constant are also visible by the other ones), while they are in case
       of opaque constants, so they are hidden inside each of them.
       As an optimization one could attach only the ones actually used in the
       opaque case. *)
    let fst_eff, more_eff =
```

Reply:
That reads great thanks!